Tutorials on building PyRenew multisignal models (#819)
CDCgov/PyRenew: docs/tutorials/multisignal_H_E_model.qmd

    delay_df = pd.DataFrame(
        {
            "days": days,
            "probability": delay_pmf,
        }
    )

    max_probability = delay_df["probability"].max()

    return (
        p9.ggplot(delay_df, p9.aes(x="days", y="probability"))
        + p9.geom_col(fill="steelblue", alpha=0.7, color="black")
        + p9.geom_vline(
            xintercept=mode_delay,
            color="purple",
            linetype="solid",
            size=1,
        )
        + p9.geom_vline(
            xintercept=mean_delay,
            color="red",
            linetype="dashed",
            size=1,
        )
        + p9.labs(
            x="Days",
            y="Probability",
            title=title,
        )
        + theme_tutorial
        + p9.annotate(
            "text",
            x=mode_delay + 8,
            y=max_probability * 0.95,
            label=f"Mode: {mode_delay} days",
            color="purple",
            size=10,
        )
        + p9.annotate(
            "text",
            x=mean_delay + 8,
            y=max_probability * 0.8,
            label=f"Mean: {mean_delay:.1f} days",
            color="red",
            size=10,
        )
    )
```

> This tutorial builds a small hospital + emergency department model using PyRenew's high-level components and `PyrenewBuilder` class to specify a `MultiSignalModel` object, then fits it to a synthetic dataset.

Reviewer: ```suggestion
This tutorial builds a small hospital (H) + emergency department (E) model using PyRenew's high-level components and `PyrenewBuilder` class to specify a `MultiSignalModel` object, then fits it to a synthetic dataset.
```

The model has one latent infection process.
Both observations receive the same aggregate infection trajectory, but they map it to data on different reporting grids.

```mermaid
flowchart TB

    subgraph Latent["Latent Infection Process"]
        R["Weekly DifferencedAR1<br/>for log Rt"]
        I["Daily renewal equation<br/>(PopulationInfections)"]
    end

    subgraph Shared["Shared Daily Axis"]
        D["Daily latent infections"]
    end

    subgraph Asc["Ascertainment"]
        A["JointAscertainment<br/>scalar IHR and IEDR"]
    end

    subgraph Obs["Observation Processes"]
        H["Hospital admissions<br/>PopulationCounts<br/>weekly MMWR totals"]
        E["ED visits<br/>PopulationCounts<br/>daily counts + day-of-week"]
    end

    subgraph Data["Observed Data"]
        HD["Weekly hospital data"]
        ED["Daily ED data"]
    end

    R -->|"broadcast to days"| I
    I --> D
    D --> H
    D --> E
    A --> H
    A --> E
    H -->|"delay + IHR + weekly aggregation + noise"| HD
    E -->|"delay + IEDR + day-of-week + noise"| ED
```

The components are:

- **Latent infections**: `PopulationInfections` produces a single aggregate infection trajectory.

> - **$\mathcal{R}(t)$process**: `WeeklyTemporalProcess` samples weekly values and broadcasts them to the daily renewal equation.

Reviewer: ```suggestion
- **$\mathcal{R}(t)$ process**: `WeeklyTemporalProcess` samples weekly values and broadcasts them to the daily renewal equation.
```

These alternatives are good candidates for model comparison and sensitivity analysis.
One advantage of specifying the model with `PyrenewBuilder` is that closely related models can be built from the same components while changing only the feature under study: fixed versus inferred delays, scalar versus time-varying ascertainment, or renewal dynamics with versus without feedback.
In practice, each candidate model can be defined as a named Python object, factory function, or separate model specification file.
This makes it easier to compare posterior predictive checks, held-out forecast performance, and the stability of inferred latent infections across model variants.

## Specifying Model Components for `PyrenewBuilder`

The `PyrenewBuilder` class assembles the model components.
It checks component compatibility and coherence, computes initialization requirements, and produces a model ready for inference.

### Latent Infection Process

The `PyrenewBuilder` object's `configure_latent` method handles configuration of the latent process.
It requires the following:

- **Generation interval**: PMF for secondary infection timing.

> - **Initial conditions ($I(0)$ and $\log(\mathcal{R}(0))$)**: Infection prevalence at time $0$ as a proportion of the population, and the starting point of the $\log(\mathcal{R}(t))$ trajectory.

Reviewer: ```suggestion
- **Initial conditions ($I(0)$ and $\log(\mathcal{R}(0))$)**: Infection incidence at time $0$ as a proportion of the population, and the starting point of the $\log(\mathcal{R}(t))$ trajectory.
```

)
```

`WeeklyTemporalProcess` uses calendar weeks, not arbitrary 7-day blocks starting at index 0.
The `start_dow=MMWR_WEEK` argument says that each $\mathcal{R}(t)$ value applies to an MMWR week, which runs Sunday through Saturday.

At model runtime, `obs_start_date` is used to align the weekly $\mathcal{R}(t)$ blocks with the daily model axis.
The same calendar anchor is used when daily predicted hospital admissions are aggregated to MMWR weeks.

### Ascertainment

`JointAscertainment` defines one shared ascertainment component for the hospital and ED signals.
The component has a `name`, used to identify it in the multisignal model, and a tuple of signal names supplied through `signals`.

For each signal, the model samples a scalar ascertainment rate.
The rates are modeled jointly: `baseline_rates` gives their prior centers on the probability scale, while `covariance_matrix` describes uncertainty and dependence on the logit scale.
The logit-scale representation keeps sampled rates between 0 and 1 after transformation back to the probability scale.

We use a common generic baseline rate and a logit-scale standard deviation of $0.3$, matching the prior scale used by custom-H-E.

> The prior correlation of $0.5$ says that the two rates may differ, but prior draws that are high for hospital ascertainment also tend to be high for ED ascertainment.

Reviewer: Mention that correlated ascertainment is important for identifiability.

Author: noted.
we could show this by introducing another model which estimates each separately.   there's a second tutorial not included in this PR on workflow which builds successive versions of the model.

Reviewer: > another model which estimates each separately

is this not what Pyrenew-h-e is?

Author: each observation process estimates the ascertainment rate without linking the two - 
cf the ascertainment tutorial section:  https://cdcgov.github.io/PyRenew/tutorials/ascertainment.html#independent-scalar-ascertainment

- **Noise model**: the count distribution used to model variation around the expected number of events.

In addition, count data observation processes have optional parameters for:

- **Day-of-week effects**: recurring multiplicative adjustments for systematic differences among days of the week.
- **Right truncation**: adjustment for recent observations that may be incomplete because not all reports have arrived.

#### ED Visits

For ED visits, we include a day-of-week effect.

ED visits often have systematic day-of-week patterns.
Day-to-day volumes may differ because of care-seeking behavior, clinic availability, staffing, coding, or reporting workflows.
Those patterns are part of the observation process, not necessarily evidence that infections changed from one day to the next.

The ED observation model includes $7$ multiplicative effects, one for each day of the week.
The day-of-week effect multiplies the expected daily ED visits before the count observation model is applied.
Each day of the week has its own multiplier: a value above 1 increases expected ED visits on that day, and a value below 1 decreases them.

> Following the custom H+E model, we place a symmetric Dirichlet prior with concentration value $5$ on these weights and scale the draw by $7$.

Reviewer: @dylanhmorris Can you suggest a justification to include here?

Author: will change to:
Following the custom H+E model, we place a symmetric Dirichlet($\alpha=5$) prior on the seven day-of-week weights and rescale by 7 so they act as multiplicative factors with mean 1,
encoding a mild prior belief that day-of-week effects are modest deviations from a flat week.

Reviewer: Is the use of Dirichlet (rather than, say, 7 independent independent gamma distributions), for identifiability purposes? It's a bit strange, in my opinion (but it is done this way in all of the previous pyrenew-esque models).

```{python}
#| code-fold: false
#| label: align-observed-data

ed_observed = model.pad_observations(
    jnp.array(daily_ed["ed_visits"].to_numpy(), dtype=jnp.float32)
)

hospital_observed = model.pad_aggregated_observations(
    jnp.array(weekly_hosp["weekly_hosp_admits"].to_numpy(), dtype=jnp.float32),
    observation_name="hospital",
    n_days_post_init=N_DAYS_FIT,
    obs_start_date=OBS_START_DATE,
)

print(f"ED observation array shape: {ed_observed.shape}")
print(f"Hospital observation array shape: {hospital_observed.shape}")
```

```{python}
#| label: plot-observed-data

# Raw observations on a shared calendar axis
ed_dates = pd.date_range(
    start=pd.Timestamp(OBS_START_DATE),
    periods=len(daily_ed),
    freq="D",
)

ed_plot_df = pd.DataFrame(
    {
        "date": ed_dates,
        "count": daily_ed["ed_visits"].to_numpy(),
        "signal": "Daily ED visits",
    }
)

hosp_plot_df = pd.DataFrame(
    {
        "week_end": pd.to_datetime(weekly_hosp["week_end"]),
        "count": weekly_hosp["weekly_hosp_admits"].to_numpy(),
        "signal": "Weekly hospital admissions",
    }
)
hosp_plot_df["week_start"] = hosp_plot_df["week_end"] - pd.Timedelta(days=6)

> raw_obs_plot = (

Reviewer: I was thinking two dot plots would be a better visualization, but I kind of like that the thickness of the bars calls attention to the daily vs weekly scale.

            "IEDR_mean": custom_draw_array("IEDR_mean"),
            "latent_infections": custom_latent_infections,
            **({"rt": custom_rt} if custom_rt is not None else {}),
        }
    },
    dims={
        "latent_infections": ["time"],
        **({"rt": ["time"]} if custom_rt is not None else {}),
    },
    sample_dims=["chain", "draw"],
)
```

We use ArviZ for convergence checks and posterior summaries.
The first `n_init` daily time points are initialization points, so the saved daily posterior quantities are trimmed before summarizing or plotting.

```{python}
#| label: arviz-checks-recovered-params

pyrenew_parameter_summary = az.summary(
    idata_trimmed,
    var_names=[
        "I0",
        "log_rt_time_0",
        "he_ascertainment_hospital",
        "he_ascertainment_ed_visits",
    ],
)
pyrenew_parameter_summary
```

```{python}
pyrenew_ratio = (
    idata_trimmed.posterior["he_ascertainment_hospital"]
    / idata_trimmed.posterior["he_ascertainment_ed_visits"]
)

az.summary(
    az.from_dict(
        {"posterior": {"IHR_over_IEDR": pyrenew_ratio.values}},
        sample_dims=["chain", "draw"],
    )
)
```

> We compare the model estimates to the data-generating parameters.

Reviewer: Can we make it easier to compare the true value to the posterior? By putting them in the same table?

Author: yes, put them together into a table.

days = np.arange(len(rt_q50)) - n_init

rt_df = pd.DataFrame(
    {
        "day": days,
        "q05": rt_q05,
        "median": rt_q50,
        "q95": rt_q95,
        "period": np.where(days < 0, "Initialization", "Observed period"),
    }
)

rt_plot = (
    p9.ggplot(rt_df, p9.aes(x="day"))
    + p9.geom_ribbon(
        p9.aes(ymin="q05", ymax="q95"),
        fill="steelblue",
        alpha=0.25,
    )
    + p9.geom_line(
        p9.aes(y="median"),
        color="steelblue",
        size=1,
    )
    + p9.geom_hline(yintercept=1.0, linetype="dashed", alpha=0.6)
    + p9.geom_vline(xintercept=0, linetype="dotted", alpha=0.8)
    + p9.annotate(
        "text",
        x=-n_init / 2,
        y=max(rt_q95) * 0.95,
        label="Initialization period",
        size=9,
        color="gray",
    )
    + p9.annotate(
        "text",
        x=20,
        y=max(rt_q95) * 0.95,
        label="Observed period",
        size=9,
        color="gray",
    )
    + p9.labs(
        x="Day relative to first observation",
        y="Rt",
        title="Posterior Rt Over Initialization and Observation Periods",
    )
    + theme_tutorial
)

rt_plot
```

> Rt is useful as a companion to latent infections because it focuses on epidemic shape rather than absolute infection scale.

Reviewer: ```suggestion
$\mathcal{R}(t)$  is useful as a companion to latent infections because it focuses on epidemic shape rather than absolute infection scale.
```

> The plot below compares Rt across all three models against the synthetic data-generating Rt trajectory.

Reviewer: The intro only mentions 2 models. Perhaps I was not reading carefully enough. Can we state what the three models are at the very beginning?

Reviewer: I think the names could be a bit clearer, as well.

Something like

pyrenew-H-E -> independent-ascertainment
linked-pyrenew-H-E -> correlated-ascertainment
custom-H-E -> ratio-ascertainment

?

            "truth": true_rt[:N_DAYS_FIT],
        }
    ),
    pd.DataFrame(
        {
            "model": "linked-pyrenew-H-E",
            "day": np.arange(N_DAYS_FIT),
            "q05": linked_rt_q.sel(quantile=0.05).values[:N_DAYS_FIT],
            "median": linked_rt_q.sel(quantile=0.5).values[:N_DAYS_FIT],
            "q95": linked_rt_q.sel(quantile=0.95).values[:N_DAYS_FIT],
            "truth": true_rt[:N_DAYS_FIT],
        }
    ),
]

if "rt" in custom_scalar_idata.posterior:
    custom_rt = custom_scalar_idata.posterior["rt"].squeeze()
    custom_rt_q = custom_rt.quantile([0.05, 0.5, 0.95], dim=["chain", "draw"])
    rt_comparison_parts.append(
        pd.DataFrame(
            {
                "model": "custom-H-E",
                "day": np.arange(N_DAYS_FIT),
                "q05": custom_rt_q.sel(quantile=0.05).values[:N_DAYS_FIT],
                "median": custom_rt_q.sel(quantile=0.5).values[:N_DAYS_FIT],
                "q95": custom_rt_q.sel(quantile=0.95).values[:N_DAYS_FIT],
                "truth": true_rt[:N_DAYS_FIT],
            }
        )
    )

rt_comparison_df = pd.concat(rt_comparison_parts, ignore_index=True)
```

> ```{python}

Reviewer: This figure is quite cramped. Override the default size?

<img width="960" height="576" alt="Image" src="https://github.com/user-attachments/assets/dfbe6b53-5693-4ee2-acd5-647c29895b38" />

Author: fixed this and several other similar plots.   the tutorial theme fonts seem too small - I made pretty much the same set of fixes to all the plots:
```
    + p9.theme(
        text=p9.element_text(size=16),
        plot_title=p9.element_text(size=18, weight="bold"),
        axis_title=p9.element_text(size=15),
        axis_text=p9.element_text(size=13),
        strip_text=p9.element_text(size=14),
```

needs more investigation.
according to the AI,  mkdocs is generating the png and which is then restricted to column width, but if the theme specifies the figure size, they should all be the same ???

For hospital and ED count observations, the expected count is approximately latent infections times an ascertainment rate, after delay convolution and calendar aggregation.
If the posterior puts IHR or IEDR below the data-generating value, the model can still match the observed counts by putting latent infections above the data-generating trajectory.

```{python}
#| label: synthetic-recovery-check

rhat = pyrenew_parameter_summary["r_hat"].astype(float)
ess = pyrenew_parameter_summary["ess_bulk"].astype(float)
print(f"Max R-hat: {rhat.max():.3f}")
print(f"Min bulk ESS: {ess.min():.0f}")

latent_inf = idata_trimmed.posterior["latent_infections"]
inf_q05 = latent_inf.quantile(0.05, dim=["chain", "draw"]).values
inf_q50 = latent_inf.quantile(0.50, dim=["chain", "draw"]).values
inf_q95 = latent_inf.quantile(0.95, dim=["chain", "draw"]).values

true_infections = daily_infections["true_infections"].to_numpy()
n_compare = min(len(true_infections), len(inf_q05))
inf_covered = (true_infections[:n_compare] >= inf_q05[:n_compare]) & (
    true_infections[:n_compare] <= inf_q95[:n_compare]
)
print(f"90% interval coverage for true infections: {np.mean(inf_covered):.1%}")

> ```

Reviewer: This table needs more commentary. I'm not sure what it's about.

ed_ma = pd.Series(ed_raw).rolling(window=7, center=True).mean().values
ed_df = pd.DataFrame(
    {
        "day": np.arange(len(ed_raw)),
        "median": ed_ma,
        "raw": ed_raw,
        "q05": np.nan,
        "q95": np.nan,
        "signal": "Daily ED Visits (7-day MA)",
    }
)

plot_df = pd.concat([infections_df, hosp_df, ed_df], ignore_index=True)
plot_df["signal"] = pd.Categorical(
    plot_df["signal"],
    categories=[
        "Weekly Hospital Admissions",
        "Daily ED Visits (7-day MA)",
        "Latent Infections",
    ],
    ordered=True,
)

(
    p9.ggplot(plot_df, p9.aes(x="day"))
    + p9.geom_ribbon(
        p9.aes(ymin="q05", ymax="q95"),
        fill="steelblue",
        alpha=0.3,
    )
    + p9.geom_point(
        p9.aes(y="raw"),
        color="gray",
        alpha=0.35,
        size=1,
    )
    + p9.geom_line(
        p9.aes(y="median"),
        color="darkblue",
        size=1,
    )
    + p9.facet_wrap("~signal", ncol=1, scales="free_y")
    + p9.scale_y_log10()
    + p9.labs(
        x="Day",
        y="Count",
        title="Posterior Latent Infections vs Observed H+E Data",
    )
    + theme_tutorial
)

> ```

Reviewer: I think it would make more sense here to show the posterior predictive for E and H, not just a simple MA.

        "day": np.arange(n_compare),
        "relative_width": inf_relative_width,
        "period": pd.cut(
            np.arange(n_compare),
            bins=period_edges,
            labels=period_labels,
        ),
    }
)

period_summary_df = (
    ci_width_df.groupby("period", observed=True)
    .agg(
        start=("day", "min"),
        end=("day", "max"),
        mean_relative_width=("relative_width", "mean"),
    )
    .reset_index()
)

print("\nMean relative 90% interval width by time period:")
for row in period_summary_df.itertuples(index=False):
    print(f"  {row.period}: {row.mean_relative_width:.2f}")

(
    p9.ggplot(ci_width_df, p9.aes(x="day", y="relative_width"))
    + p9.geom_line(color="gray", alpha=0.6, size=0.7)
    + p9.geom_segment(
        data=period_summary_df,
        mapping=p9.aes(
            x="start",
            xend="end",
            y="mean_relative_width",
            yend="mean_relative_width",
            color="period",
        ),
        size=1.6,
    )
    + p9.labs(
        x="Day",
        y="Relative 90% interval width",
        color="Period",
        title="Posterior Uncertainty in Latent Infections",
    )
    + theme_tutorial
)

> ```

Reviewer: What do we learn from this plot that we didn't learn from the previous one? Why bin into two week estimates?

Reviewer: After further reading, I see that this introduces the idea of relative widths, which are an important comparison point between the models later in the document.

    [
        pd.DataFrame(
            {
                "model": "pyrenew-H-E",
                "day": np.arange(n_compare),
                "relative_width": inf_relative_width[:n_compare],
            }
        ),
        pd.DataFrame(
            {
                "model": "linked-pyrenew-H-E",
                "day": np.arange(n_compare),
                "relative_width": (
                    linked_inf_q95[:n_compare] - linked_inf_q05[:n_compare]
                )
                / np.maximum(linked_inf_q50[:n_compare], 1.0),
            }
        ),
        pd.DataFrame(
            {
                "model": "custom-H-E",
                "day": np.arange(n_compare),
                "relative_width": (
                    custom_inf_q95[:n_compare] - custom_inf_q05[:n_compare]
                )
                / np.maximum(custom_inf_q50[:n_compare], 1.0),
            }
        ),
    ],
    ignore_index=True,
)

latent_width_summary = (
    latent_width_df.groupby("model", observed=True)
    .agg(
        mean_relative_width=("relative_width", "mean"),
        median_relative_width=("relative_width", "median"),
    )
    .reset_index()
)
latent_width_summary

> ```

Reviewer: The above table needs much more detail, I think. Mean relative width of what?

```{python}
#| label: linked-latent-infections

> #| fig-cap: Posterior latent infections under the two PyRenew parameterizations

Reviewer: I just realized these figure captions aren't showing in the html (I have just been reviewing the html, not the qmd). We should figure out how to display them.

Author: another glitch in the toolchain.   made all captions into italicized paragraph directly under the plot.  not elegant, but it works.

Reviewer: See https://github.com/CDCgov/PyRenew/issues/826. Let's decide how to address this repo-wide before merging this PR.

#|   and custom-H-E. The dashed black line is the synthetic data-generating
#|   latent infection trajectory, not an observed signal used for fitting.

(
    p9.ggplot(latent_comparison_df, p9.aes(x="day"))
    + p9.geom_ribbon(
        p9.aes(ymin="q05", ymax="q95"),
        fill="steelblue",
        alpha=0.25,
    )
    + p9.geom_line(p9.aes(y="median"), color="darkblue", size=1)
    + p9.geom_line(p9.aes(y="truth"), color="black", linetype="dashed", size=0.7)
    + p9.facet_wrap("~model", ncol=1)
    + p9.scale_y_log10()
    + p9.labs(
        x="Day",
        y="Latent infections",
        title="Posterior Latent Infections vs Synthetic Truth",
    )
    + theme_tutorial
)

> ```

Reviewer: A bit cramped. Can we make it taller? 

<img width="960" height="576" alt="Image" src="https://github.com/user-attachments/assets/86a05391-1575-4e02-93b9-bff148c730cf" />

Author: yes, made it taller.

                "ratio": custom_ihr / custom_iedr,
            }
        ),
    ],
    ignore_index=True,
)

scale_density_df = scale_draws_df.melt(
    id_vars="model",
    value_vars=["I0", "IHR", "IEDR"],
    var_name="quantity",
    value_name="value",
)

truth_scale_df = pd.DataFrame(
    {
        "quantity": ["I0", "IHR", "IEDR"],
        "truth": [true_i0, true_ihr, true_iedr],
    }
)
```

The marginal densities show how each model places posterior mass for the absolute-scale quantities.

```{python}
#| label: identifiability-scale-densities
#| fig-cap: |-
#|   Posterior densities for $I_0$, IHR, and IEDR under all three models.
#|   Red dashed lines mark data-generating values.

(
    p9.ggplot(scale_density_df, p9.aes(x="value", fill="model", color="model"))
    + p9.geom_density(alpha=0.35)
    + p9.geom_vline(
        truth_scale_df,
        p9.aes(xintercept="truth"),
        color="red",
        linetype="dashed",
    )
    + p9.facet_wrap("~quantity", scales="free")
    + p9.scale_x_log10()
    + p9.labs(
        x="Value",
        y="Density",
        fill="Model",
        color="Model",
        title="Posterior Scale Parameters",
    )
    + theme_tutorial
)

> ```

Reviewer: yikes. They all fail to recover the IEDR and IHR. Can we plot the priors too? Perhaps in a facet above?

The joint posterior for IHR and IEDR shows how each model constrains the two clinical ascertainment rates together.
The red point marks the data-generating values.

```{python}
#| label: identifiability-ihr-iedr
#| fig-cap: |-
#|   Joint posterior of hospital and ED ascertainment rates under both
#|   PyRenew parameterizations and custom-H-E.

truth_asc_df = pd.DataFrame({"IHR": [true_ihr], "IEDR": [true_iedr]})

(
    p9.ggplot(scale_draws_df, p9.aes(x="IHR", y="IEDR", color="model"))
    + p9.geom_point(alpha=0.16, size=0.7)
    + p9.geom_point(data=truth_asc_df, color="red", size=3)
    + p9.facet_wrap("~model")
    + p9.scale_x_log10()
    + p9.scale_y_log10()
    + p9.labs(
        x="Hospital ascertainment (IHR)",
        y="ED ascertainment (IEDR)",
        color="Model",
        title="Hospital and ED Ascertainment Are Jointly Constrained",
    )
    + theme_tutorial
)

> ```

Reviewer: Would love to see priors here too.
